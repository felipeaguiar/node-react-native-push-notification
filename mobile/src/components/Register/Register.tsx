import React from 'react';
import { View, TextInput, Button, Text, StyleSheet } from 'react-native';
import AsyncStorage from '@react-native-community/async-storage';
import { Formik } from 'formik';
import * as yup from 'yup';
import { useDispatch } from 'react-redux';
import { savaTokenAction, failAction } from '../../store/ducks/auth/actions';
import api, { isApiError } from '../../api';
import { jwtUtil } from '../../util/jwt';

export default function Register() {
  const dispatch = useDispatch();

  const initialValues = {
    nome: '',
    email: '',
    senha: '',
    confirmar: ''
  };

  const validationSchema = yup.object().shape({
    nome: yup.string()
      .min(4, 'Nome inválido')
      .required('Nome é um campo obrigatório.'),
    email: yup.string()
      .email('Email invalido')
      .required('E-mail é um campo obrigatório.'),
    senha: yup.string()
      .min(6, 'Senha deve ter no mínimo 6 caracteres')
      .required('Senha é um campo obrigatório.'),
    confirmar: yup.string()
      .min(6, 'Confirmar Senha deve ter no mínimo 6 caracteres')
      .oneOf([yup.ref('senha')], 'Senha e Confirmação de senha não podem ser diferentes')
      .required('Confirmar Senha é um campo obrigatório.')
  });

  async function onSubmit(values: typeof initialValues) {
    try {
      const response = await api.post('auth/register', {
        nome: values.nome,
        email: values.email,
        senha: values.senha
      });
      values.senha = '';
      values.confirmar = '';

      const token = response.data.token;
      AsyncStorage.setItem('token', token);

      const jwt = jwtUtil.decode(token);

      if (!jwt || !jwtUtil.isExpired(jwt)) {
        throw new Error();
      }

      dispatch(savaTokenAction(token, jwt));
    } catch (error) {
      let message = 'Erro ao efetuar login';

      if (isApiError(error)) {
        message = error.response.data.errors[0].message;
      }

      await AsyncStorage.removeItem('token');
      dispatch(failAction(message));
    }
  }

  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema}>
        {({ handleChange, handleBlur, handleSubmit, values, errors, touched, isSubmitting }) => (
        <View style={styles.container}>

          <TextInput
            placeholder="Nome Completo"
            autoCapitalize="none"
            onChangeText={handleChange('nome')}
            onBlur={handleBlur('nome')}
            value={values.nome}
          />

          { errors.nome && touched.nome && <Text>{errors.nome}</Text> }

          <TextInput
            placeholder="Endereço de E-mail"
            autoCapitalize="none"
            onChangeText={handleChange('email')}
            onBlur={handleBlur('email')}
            value={values.email}
          />

          { errors.email && touched.email && <Text>{errors.email}</Text> }

          <TextInput
            placeholder="Senha"
            secureTextEntry
            onChangeText={handleChange('senha')}
            onBlur={handleBlur('senha')}
            value={values.senha}
          />

          { errors.senha && touched.senha && <Text>{errors.senha}</Text> }

          <TextInput
            placeholder="Confirmar Senha"
            secureTextEntry
            onChangeText={handleChange('confirmar')}
            onBlur={handleBlur('confirmar')}
            value={values.confirmar}
          />

          { errors.confirmar && touched.confirmar && <Text>{errors.confirmar}</Text> }

          <Button title="Criar Conta" disabled={isSubmitting} onPress={handleSubmit} />
        </View>
      )}
    </Formik>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  }
});
