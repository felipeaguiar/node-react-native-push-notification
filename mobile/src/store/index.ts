import { createStore, combineReducers} from 'redux';
import { authReducer } from './ducks/auth/reducer';

export const rootReducer = combineReducers({
  auth: authReducer
});

const store = createStore(rootReducer);

export type RootState = ReturnType<typeof rootReducer>
export type Store = typeof store;

export default store;
