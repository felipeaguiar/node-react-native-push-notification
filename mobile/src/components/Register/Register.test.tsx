import '@testing-library/jest-native/extend-expect';
import React from 'react';
import { fireEvent, waitFor, act } from 'react-native-testing-library';
import AsyncStorage from '@react-native-community/async-storage';
import Register from './Register';
import { registerMock } from '../../test/http-mock';
import store from '../../store';
import { renderWithRedux } from '../../test/reduxMock';

it('register snapshot', () => {
  expect.hasAssertions();

  const { toJSON } = renderWithRedux(<Register />);
  expect(toJSON()).toMatchSnapshot();
});

it('registrando com sucesso', async () => {
  expect.hasAssertions();

  registerMock();
  const { toJSON, getByPlaceholder, getByText } = renderWithRedux(<Register />, { store } );

  const nome = getByPlaceholder('Nome Completo');
  const email = getByPlaceholder('Endereço de E-mail');
  const senha = getByPlaceholder('Senha');
  const confirmar = getByPlaceholder('Confirmar Senha');
  const criar = getByText('Criar Conta');

  fireEvent.changeText(nome, 'Felipe Aguiar');
  fireEvent.changeText(email, 'felipe@mail.com');
  fireEvent.changeText(senha, '123456');
  fireEvent.changeText(confirmar, '123456');

  fireEvent.press(criar);

  await waitFor(() => expect(criar).toBeEnabled());

  const token = await AsyncStorage.getItem('token');

  expect(nome).toHaveProp('value', expect.any(String));
  expect(email).toHaveProp('value', expect.any(String));

  expect(senha).toHaveProp('value', '');
  expect(confirmar).toHaveProp('value', '');

  expect(token).not.toBeNull();
  expect(token).toBe(store.getState().auth.token);

  expect(toJSON()).toMatchSnapshot();
});

it('registrando com falha', async () => {
  expect.hasAssertions();

  registerMock();
  const { toJSON, getByPlaceholder, getByText, queryByText } = renderWithRedux(<Register />, { store } );

  const nome = getByPlaceholder('Nome Completo');
  const email = getByPlaceholder('Endereço de E-mail');
  const senha = getByPlaceholder('Senha');
  const confirmar = getByPlaceholder('Confirmar Senha');
  const criar = getByText('Criar Conta');

  fireEvent.changeText(nome, 'Fel');
  fireEvent.changeText(email, 'mail.com');
  fireEvent.changeText(senha, '123');
  fireEvent.changeText(confirmar, '654321');

  await act(async() => fireEvent.press(criar));

  const nomeError = queryByText('Nome inválido');
  const emailError = queryByText('Email invalido');
  const senhaError = queryByText('Senha deve ter no mínimo 6 caracteres');
  const confirmarError = queryByText('Senha e Confirmação de senha não podem ser diferentes');

  expect(nomeError).toBeTruthy();
  expect(emailError).toBeTruthy();
  expect(senhaError).toBeTruthy();
  expect(confirmarError).toBeTruthy();

  expect(nome).toHaveProp('value', expect.any(String));
  expect(email).toHaveProp('value', expect.any(String));

  expect(toJSON()).toMatchSnapshot();
});
