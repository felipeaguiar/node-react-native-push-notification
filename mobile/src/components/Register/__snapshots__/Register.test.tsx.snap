// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Nome Completo"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Endereço de E-mail"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Confirmar Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Criar Conta
      </Text>
    </View>
  </View>
</View>
`;

exports[`registrando com falha 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Nome Completo"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value="Fel"
  />
  <Text>
    Nome inválido
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Endereço de E-mail"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value="mail.com"
  />
  <Text>
    Email invalido
  </Text>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value="123"
  />
  <Text>
    Senha deve ter no mínimo 6 caracteres
  </Text>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Confirmar Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value="654321"
  />
  <Text>
    Senha e Confirmação de senha não podem ser diferentes
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Criar Conta
      </Text>
    </View>
  </View>
</View>
`;

exports[`registrando com sucesso 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Nome Completo"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value="Felipe Aguiar"
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Endereço de E-mail"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value="felipe@mail.com"
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Confirmar Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Criar Conta
      </Text>
    </View>
  </View>
</View>
`;
